src/QirRuntime/lib/QIR/conditionals.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include <stdexcept>
+
+#include "quantum__qis.hpp"
+
+#include "QirTypes.hpp"
+#include "quantum__rt.hpp"
+
+static void ApplyWithFunctor(bool isControlled, bool isAdjoint, QirArray* ctls, QirCallable* clb)
+{
+    auto tupleSize = isControlled ? sizeof(void*) : 0;
+    PTuple argsTuple = quantum__rt__tuple_create(tupleSize);
+
+    QirCallable* clbFunc =
+        (isAdjoint || isControlled) ? quantum__rt__callable_copy(clb, true /*force new instance*/) : clb;
+
+    if (isAdjoint)
+    {
+        quantum__rt__callable_make_adjoint(clbFunc);
+    }
+    if (isControlled)
+    {
+        quantum__rt__callable_make_controlled(clbFunc);
+        *reinterpret_cast<QirArray**>(argsTuple) = ctls;
+    }
+
+    quantum__rt__callable_invoke(clbFunc, argsTuple /*args*/, nullptr /*result*/);
+
+    if (clb != clbFunc)
+    {
+        quantum__rt__callable_update_reference_count(clbFunc, -1);
+    }
+    quantum__rt__tuple_update_reference_count(argsTuple, -1);
+}
+
+static bool ArraysContainEqualResults(QirArray* rs1, QirArray* rs2)
+{
+    assert(rs1 != nullptr && rs2 != nullptr && rs1->count == rs2->count);

```

Comment:
As for `assert( .. , rs1->count == rs2->count)` (comparing the array counts). 
I assume this fragment is only called by _our code_ (compiler generated or written manually by us, etc.), the _user_ never calls this fragment (directly or indirectly). And I assume that if the comparison/assertion fails then it definitely means a bug in our code. That is why it is the right thing to assert if the arrays have different count.
(But if the _user_ can call this fragment (directly or indirectly) and pass the arrays with different count then I'm not sure if assertion is the right thing to do)

Reply:
It's a good question of at what layer of the code (Q#, QIR, QirRuntime) checks like this should be done. I can easily replace the assert with fail, but I don't think that "defensive" coding is _always_ the right answer. We should think through where we can draw the boundaries that would allow us to assume the inputs has been already checked... For now, I'll keep the checks as asserts.